### PR TITLE
fix(monitoring): isolate time series by request project

### PR DESCRIPTION
**Pull request by vedit**
## Summary

Scope Monitoring storage access to the project named in each request, preventing time-series data from leaking across projects on gRPC calls. Add explicit project operations to the existing project-aware backend. This is a three-file fix, extracted from #191 at the maintainer's request.

## Type of change

- [x] Bug fix (`fix:`)
- [ ] New feature (`feat:`)
- [ ] Breaking change (`feat!:`)
- [ ] Docs / chore

## GCP Compatibility

56 focused Monitoring/storage tests passed. The official Java Monitoring 3.97.0 synthetic-counter fixture also passed against this independent emulator branch; that fixture is proposed in the separate contracts PR.

[Reconstructed-set JVM/native evidence](https://github.com/vedit/floci-gcp/actions/runs/34859974829), source `df57d04c5c33cd144d9b1111fd49021b240e178a`. Combined evidence is distinct from the focused standalone checks above. All branches start from upstream `1bd6583`; no newer upstream changes are reverted.

## Checklist

- [x] Focused validation completed as described above
- [x] New or updated automated coverage included or linked to the dependent contracts contribution
- [x] Commit messages follow Conventional Commits

## Greptile follow-up

Agreed. Added direct service-level coverage in 84f61b4 using ProjectAwareStorageBackend for both descriptor and time-series stores, with an ambient project different from either request project.

The regression writes the same metric identity and timestamp in two projects, then checks descriptor get/list, per-project time-series values, and descriptor deletion with cascading point removal. Deleting one project preserves the other, and neither writes into the ambient project's store. Monitoring does not expose a separate time-series get/delete method; series are retrieved with list and deleted here through descriptor deletion.

Validation: 52 Monitoring/storage tests passed with no failures or skips. The negative control replaces the request-bound scope with the old ambient store: the new test fails because the second project's point collides with the first project's timestamp. This coverage is now in this standalone PR rather than depending on the later SDK-contracts PR.

## Maintainer follow-up (2026-09-21)

Fixed in a6f9103 and rebased onto current main (206719d).

ListTimeSeries now accepts organization and folder names over gRPC. With no hierarchy membership model, those reads return an empty page without a continuation token; they never scan another project's metrics. Filter, interval and aggregation validation still run. Descriptor operations and time-series writes remain project-only.

Added service coverage with data in two projects plus a generated MetricServiceGrpc client regression for organizations/folders in FULL and HEADERS views. Invalid filter/interval and malformed names remain rejected. The docs describe the empty-result limitation explicitly.

Validation: all 59 focused Monitoring/storage tests passed, no failures or skips. The contract is based on the cited [Monitoring v3 proto](https://github.com/googleapis/googleapis/blob/aa87617d6732d348ff158d20d474c0dfba3d0086/google/monitoring/v3/metric_service.proto#L368-L374); no live hierarchy semantics are claimed.

These are fresh focused JVM/SDK results; the full native matrix linked above predates this rebase. Upstream checks on the updated head are tracked separately.

### CODE REVIEW

File: src/main/java/io/floci/gcp/services/cloudmonitoring/CloudMonitoringService.java [hunk 1]
```diff
@@ -120,6 +120,14 @@ void onStart(@Observes StartupEvent ev) {
         grpcServerManager.bind(new CloudMonitoringController(this));
     }
 
+    private static <T> StorageBackend<String, T> scoped(StorageBackend<String, T> store, String resourceName) {
+        String[] parts = resourceName.split("/", -1);
+        if (parts.length < 2 || !parts[0].equals("projects") || parts[1].isBlank()) {
+            throw GcpException.invalidArgument("A projects/{project} resource name is required");
+        }
+        return store instanceof ProjectAwareStorageBackend<T> aware ? aware.forProject(parts[1]) : store;
```
[the remark this answers is not in the document]
Comment (vedit) [reply src/main/java/io/floci/gcp/services/cloudmonitoring/CloudMonitoringService.java:128]: Agreed. Added direct service-level coverage in 84f61b4 using ProjectAwareStorageBackend for both descriptor and time-series stores, with an ambient project different from either request project.

The regression writes the same metric identity and timestamp in two projects, then checks descriptor get/list, per-project time-series values, and descriptor deletion with cascading point removal. Deleting one project preserves the other, and neither writes into the ambient project's store. Monitoring does not expose a separate time-series get/delete method; series are retrieved with list and deleted here through descriptor deletion.

Validation: 52 Monitoring/storage tests passed with no failures or skips. The negative control replaces the request-bound scope with the old ambient store: the new test fails because the second project's point collides with the first project's timestamp. This coverage is now in this standalone PR rather than depending on the later SDK-contracts PR.

**Comment by hectorvent** [comment]
Thanks for splitting this out, it made the change easy to follow, and the ambient-project test is a nice touch.

1. **(blocking)** `scoped()` now throws INVALID_ARGUMENT for any name that is not `projects/...`, which is right for five of the six methods. ListTimeSeries is the exception: its `name` also accepts `organizations/[ORGANIZATION_ID]` and `folders/[FOLDER_ID]` ([metric_service.proto](https://github.com/googleapis/googleapis/blob/aa87617d6732d348ff158d20d474c0dfba3d0086/google/monitoring/v3/metric_service.proto#L368-L374)), and those were served before this change. Would you be open to letting those two through for ListTimeSeries, for example with an empty or all-project result since the emulator does not model the hierarchy? No existing test pins the current behaviour.

I checked the leak itself: gRPC calls never populate the request project, so this fix is the right one, and `forProject().clear()` staying project-local is exactly right. Locally 57/0. Apart from the point above, no blockers from my side.

**Comment by vedit** [comment]
Fixed in a6f9103 and rebased onto current main (206719d).

ListTimeSeries now accepts organization and folder names over gRPC. With no hierarchy membership model, those reads return an empty page without a continuation token; they never scan another project's metrics. Filter, interval and aggregation validation still run. Descriptor operations and time-series writes remain project-only.

Added service coverage with data in two projects plus a generated MetricServiceGrpc client regression for organizations/folders in FULL and HEADERS views. Invalid filter/interval and malformed names remain rejected. The docs describe the empty-result limitation explicitly.

Validation: all 59 focused Monitoring/storage tests passed, no failures or skips. The contract is based on the cited [Monitoring v3 proto](https://github.com/googleapis/googleapis/blob/aa87617d6732d348ff158d20d474c0dfba3d0086/google/monitoring/v3/metric_service.proto#L368-L374); no live hierarchy semantics are claimed.

**Review by hectorvent — APPROVED** [review]
Thanks, this is exactly the right shape.

I checked the proto lines you cited and they say what you say they do, so accepting organization and folder parents is correct and the old INVALID_ARGUMENT was the bug. I also like that the service test writes into two projects first and then asserts the hierarchy read is empty, because that proves non-leakage rather than restating the implementation, and the generated stub test covers both views.

I have released the gated workflow runs since, and CI is green on a6f9103 across all thirteen checks.

Worth noting the empty page is not something the proto or the discovery document settles, so I am taking it as the defensible reading rather than a verified one. No blockers from me.
